src/rt/minfo.d

```diff
@@ -62,174 +62,382 @@ struct ModuleGroup
+            {
+                if(bt(reachable, i))
+                {
+                    // Since relevant dependencies are already marked as done
+                    // from recursion above, no reason to check for relevance,
+                    // that is a wasted op.
+                    bts(ctordone, i);
+                }
+            }
+
+            // add this module to the construction order list
+            ctors[ctoridx++] = current;
+        }
+
+        immutable(ModuleInfo)*[]
+            doSort(int function(immutable ModuleInfo *) getf)
+        {
+            clearFlags(relevant);
+            clearFlags(ctorstart);
+            clearFlags(ctordone);
+
+            // pre-allocate enough space to hold all modules.
+            ctors = (cast(immutable(ModuleInfo)**).malloc(len * (void*).sizeof))[0 .. len];
+            ctoridx = 0;
+            foreach (int idx, m; _modules)
+            {
+                // TODO: Should null ModuleInfo be allowed?
+                if (m is null) continue;
+                auto flag = getf(m);
+                if(flag & MIctor)
+                {
+                    if(flag & MIstandalone)
+                    {
+                        // can run at any time. Just run it first.
+                        ctors[ctoridx++] = m;
+                    }
+                    else
+                    {
+                        bts(relevant, idx);
                     }
                 }
+            }
 
-                if (stackidx)
-                {   // pop old value from stack
-                    --stackidx;
-                    mods    = stack[stackidx]._mods;
-                    idx     = stack[stackidx]._idx;
-                    auto m  = mods[idx++];
-                    immutable bitnum = findModule(m);
-                    assert(bitnum >= 0);
-                    if (m.flags & mask && !bts(ctordone, bitnum))
-                        ctors[cidx++] = m;
+            // now run the algorithm in the relevant ones
+            foreach (int idx, m; _modules)
+            {
+                if(bt(relevant, idx))
+                {
+                    if(!bt(ctordone, idx))
+                        processMod(idx);
                 }
-                else // done
-                    break;
             }
-            // store final number and shrink array
-            ctors = (cast(immutable(ModuleInfo)**).realloc(ctors.ptr, cidx * size_t.sizeof))[0 .. cidx];
+
+            return ctors[0 .. ctoridx];
         }
 
-        /* Do two passes: ctor/dtor, tlsctor/tlsdtor
-         */
-        sort(_ctors, MIctor | MIdtor);
-        sort(_tlsctors, MItlsctor | MItlsdtor);
+        // finally, do the sorting for both shared and tls ctors.
+        _ctors = doSort((immutable ModuleInfo * m) => (m.flags & MIstandalone) | ((m.dtor || m.ctor)? MIctor : 0));
+        _tlsctors = doSort((immutable ModuleInfo * m) => (m.flags & MIstandalone) | ((m.tlsdtor || m.tlsctor)? MIctor : 0));
```

Comment:
I was working based on my original code. I didn't realize there were flags that were now set by the compiler.